### PR TITLE
Engines with `E_NO_POWER_DECAY` flag lack engine power penalty

### PR DESCRIPTION
**Pull request by chaosvolt**
#### Summary

SUMMARY: Balance "Diminishing returns on engine power can now be ignored by parts with the E_NO_POWER_DECAY flag"

#### Purpose of change

It was mused upon a while ago by Kheir that it'd be nice if there was a way to enable an engine/motor to not count for diminishing returns. I'd assumed that the flag used to control the mechanics skill penalty for install also affected this, but no. There will always be diminishing returns.

#### Describe the solution

1. In vehicle.cpp, changed `vehicle::total_power_w` so that the count of how many engines have been found (with `cnt` only being used in the if statement that modifies effective total power at the very end) only adds engines to the list that have the new `E_NO_POWER_DECAY` flag.
2. Added `E_NO_POWER_DECAY` to flag JSON.
3. Updated JSON_FLAGS.md to document this new flag.

In vanilla, currently no engines use this flag (for now, see alternatives), but Arcana and other mods with engines designed to be stacked will benefit considerably from this.

#### Describe alternatives you've considered

Are there any standard motors or engines in vanilla we feel deserve this feature? If so would we lean in favor of more efficiency for small engines as an incentive to stack motors the player wouldn't normally care for, as a way to buff big power-hungry engines, or for making a specific type (gasoline, diesel, electric, V engine, inline, turbine, steam?, ???) of engine usable with full efficiency?

Kheir brought up a point: gameplay-wise and ignoring realism, the main benefit of the diminishing returns is to nudge the player to seek out better engines to upgrade a deathmobile instead of brute-forcing it with a million chainsaw engines or what-have-you. Based on this, the best candidates for this flag will probably by the top-tier, most powerful engines where there aren't really any options to upgrade any further. These will already have difficulties due to being power-hungry to begin with and weight will make it hard to stack them anyway, but at that point it might be reasonable to reward a player who's undertaken the commitment to make basically Ekarus' R-MCC and has run out of options for engine upgrades.

Also alternatively could just merge this flag with the behavior of `E_HIGHER_SKILL` like I first started to but then I'd need to fix vehicle efficiency test since turns out foot pedals lack this flag, so power of electric bikes is slightly different.

#### Testing

1. First, spawned an electric car and debugged on a second engine in a compiled test build prior to implementing this change.
2. Started it up and listed max speeds were 306/317 km per hour with 30 km per second acceleration.
4. Saved game and exited to implement changes.
5. Compiled and load-tested the save.
6. Walked off and back on, restarted car, confirmed listed speeds were still 306/317 km/h and 30 km/s because hadn't yet deleted the flag.
7. Tested adding the `E_NO_POWER_DECAY` flag to the motor abstract before loading in, faffed about and listed speed stats were now 331/343 km/h and 35 km/s.
8. Checked affected file for astyle.

### CODE REVIEW

File: doc/JSON_FLAGS.md [hunk 1]
```diff
@@ -1374,6 +1374,7 @@ Those flags are added by the game code to specific items (that specific welder,
 - ```E_COMBUSTION``` Is an engine that burns its fuel and can backfire or explode when damaged.
 - ```E_HEATER``` Is an engine and has a heater to warm internal vehicle items when on.
 - ```E_HIGHER_SKILL``` Is an engine that is more difficult to install as more engines are installed.
+- ```E_HIGHER_SKILL``` Engines with this flag do not affect total vehicle power suffering diminishing returns.
```
Comment (scarf005) [inline doc/JSON_FLAGS.md:1377]:
```suggestion
- ```E_NO_POWER_DECAY``` Engines with this flag do not affect total vehicle power suffering diminishing returns.
```
i think you meant `E_NO_POWER_DECAY` here.

**Review by scarf005 — CHANGES_REQUESTED** [review]

**Review by scarf005 — APPROVED** [review]
LGTM

**Comment by chaosvolt** [comment]
Any thoughts on what parts in vanilla, if any, to apply them to? Can always do it in a follow-up so.

**Comment by scarf005** [comment]
maybe engines with huge power output, like `"power": 100,000`?

**Comment by chaosvolt** [comment]
Seems reasonable yeah. I figured mainly the super motor, enhanced motor, and turbines presumably. Can always add it as a followup though.

**Review by scarf005 — CHANGES_REQUESTED** [review]
oh wait...

image
1. spawned electric car and installed 1 more large electric motor. activated both.

engine: 2 large electric motor
safe/top speed: 246/255 km/h, acceleration: 37 km/h/s

Cataclysm: Bright Nights - cbn-experimental-2023-04-28-1853-5-g739aa8ca9293_02
2. spawned large two electric with `E_NO_POWER_DECAY` added by using `[F] Flag`

Cataclysm: Bright Nights - cbn-experimental-2023-04-28-1853-5-g739aa8ca9293_03
3. made sure to install the flagged motors.

image
4. the engine power penalty is still applied.
engine: 2 large electric motor (`E_NO_POWER_DECAY`)
safe/top speed: 245/254 km/h, acceleration: 33 km/h/s

i'm not sure why this is happening, the code looked ok.

**Comment by chaosvolt** [comment]
Those are flags being slapped on the item, most likely not applying to the vehiclepart itself. I had to test the flag by editing the JSON.

**Comment by scarf005** [comment]
weird, another TODO to add to bug report.

**Comment by chaosvolt** [comment]
To be fair even if that worked I doubt that'd transfer the flag from the item to the vehiclepart...

**Review by scarf005 — APPROVED** [review]
image

used the same method and got much higher result. we do need to fix that flag thingy someday tho

**Pull request merged**